mmaprototype: clarify maxRangeMoveCount and maxLeaseTransferCount

tbg opened the pull request:
Informs https://github.com/cockroachdb/cockroach/issues/157757.

Epic: CRDB-55052

tbg commented on pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go:
```diff
@@ -272,6 +288,9 @@ func (re *rebalanceEnv) rebalanceReplicas(
 	loadDim := topKRanges.dim
 	for i := 0; i < n; i++ {
+		if re.rangeMoveCount >= re.maxRangeMoveCount {
+			return
```
the caller logs so this is ok.

tbg commented on pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go:
```diff
@@ -423,25 +442,17 @@ func (re *rebalanceEnv) rebalanceReplicas(
-	if doneShedding {
-		log.KvDistribution.VInfof(ctx, 2, "store s%d is done shedding, moving to next store", store.StoreID)
-		return
 	}
```
The loop now returns proactively, so if we end up at the end of the method we just didn't hit any limits and are returning normally.

tbg reviewed (commented):
More comments.

sumeerbhola reviewed (changes requested):
> @sumeerbhola reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157820)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 240 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe2B0KnBRfvdiKdryf6:-Oe2B0KnBRfvdiKdryf7:b-89npta) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1de6aa680a059071093b5f27a0f2a989ddafc80/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L240)):*
> does it make sense that rebalanceStores is called on behalf of a particular store

I think it makes sense from the perspective that each store desires to be able to work concurrently wrt other stores, which is correctly captured by having separate store rebalancers etc.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe29PnSBJuhEXTU0To6:-Oe29PnSBJuhEXTU0To7:b-nj16w6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1de6aa680a059071093b5f27a0f2a989ddafc80/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L30)):*
> ```Go
> // iterate over a slice of shedding stores - these are not necessarily local
> // but are instead stores which are overloaded and need to shed load, and for
> // which the particular calling local store has at least some leases (i.e. is
> ```

nit: ... may have some leases ...

I think we compute the overloaded stores without caring about whether they have any ranges for which this local store has a lease. This is harmless.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe2A0lY-xeYF7GNkt01:-Oe2A0lY-xeYF7GNkt02:b-slvfw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1de6aa680a059071093b5f27a0f2a989ddafc80/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L73)):*
> ```Go
> 
> // The caller has a fixed concurrency limit it can move ranges at, when it
> // is the sender of the snapshot. So we don't want to create too many
> ```

btw, I think the "fixed concurrency limit" comment was written prior to me realizing that the caller (the store rebalancer) is synchronous in making one change at a time. Which is another reason to only return one change, since it is likely to take at least a few seconds to make this change (so we can then afford to call ComputeChanges again, from a cost perspective).
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 197 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe2AiNm97ZNUGbKExg8:-Oe2AiNm97ZNUGbKExg9:b48bf16) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1de6aa680a059071093b5f27a0f2a989ddafc80/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L197)):*
> ```Go
> 	re.scratch.stores = map[roachpb.StoreID]struct{}{}
> 	for _, store := range sheddingStores {
> 		if re.rangeMoveCount >= re.maxRangeMoveCount {
> ```

why is this not looking at the maxLeaseTransferCount?
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 630 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe2CuQ0CyQBq-4Xq-w3:-Oe2CuQ0CyQBq-4Xq-w4:bb5lpum) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1de6aa680a059071093b5f27a0f2a989ddafc80/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L630)):*
> ```Go
> 		}
> 	}
> 	return false
> ```

What if we've created one or more lease transfers but not returned in the loop because of not hitting the max-fraction-pending or max-lease-transfer limit? Will we incorrectly return false here?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
I rebased and pushed a commit. I'll add another commit with testing once I've got that one ready.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-:-OeVleQK9Z6cT59rcL4P:by8o89t)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe2A0lY-xeYF7GNkt01:-OeVbQwv8mZ-1FxsfcaG:bb71pro) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1de6aa680a059071093b5f27a0f2a989ddafc80/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L73)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

btw, I think the "fixed concurrency limit" comment was written prior to me realizing that the caller (the store rebalancer) is synchronous in making one change at a time. Which is another reason to only return one change, since it is likely to take at least a few seconds to make this change (so we can then afford to call ComputeChanges again, from a cost perspective).
</blockquote></details>

PTAL
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 197 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe2AiNm97ZNUGbKExg8:-OeVc7UgD3OIRDvFLrYC:bmzgzsb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1de6aa680a059071093b5f27a0f2a989ddafc80/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L197)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why is this not looking at the maxLeaseTransferCount?
</blockquote></details>

This is `rebalanceReplicas`, which does store moves only. Leases can be shed only from the local store, so there's no loop that calls it multiple times - it's only called for the (up to one) `rebalanceStores` invocation for the localStoreID.

I made some updates (not exactly related to this comment) to the early return flow in an attempt to make it more transparent.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 240 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe2B0KnBRfvdiKdryf6:-OeVfOp0B4MroS9SLQ6G:bggkj23) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1de6aa680a059071093b5f27a0f2a989ddafc80/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L240)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> does it make sense that rebalanceStores is called on behalf of a particular store

I think it makes sense from the perspective that each store desires to be able to work concurrently wrt other stores, which is correctly captured by having separate store rebalancers etc.
</blockquote></details>

Ack, updated the comment.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 630 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe2CuQ0CyQBq-4Xq-w3:-OeVf_rM4ppsh7_lOVFb:b-jvap5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1de6aa680a059071093b5f27a0f2a989ddafc80/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L630)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

What if we've created one or more lease transfers but not returned in the loop because of not hitting the max-fraction-pending or max-lease-transfer limit? Will we incorrectly return false here?
</blockquote></details>

I think I wasn't exactly clear how this was supposed to work, but understand it now: if we hit a limit (either pending frac or absolute number of leases) there may be more leases and we want to skip moving replicas. I cleaned this up some. We now `return true` exactly when hitting a limit, and the caller skips transferring replicas in exactly that case.
If we emit, say, one lease transfer but don't hit a limit, we DO want to try moving replicas, right? Because there aren't any more leases we'd be willing to try anyway. Your comment suggests that you think we should skip replica moves more aggressively (when there are any lease moves, regardless of hitting limits) but I don't understand that heuristic. I might be missing something.

I'm not fully decided on what is the right thing to do if we skipped ranges because it's too soon since a failed change or because there are pending changes but it seems sensible to not try to be too clever - if these ranges are "busy" it seems sensible to not bank on them helping out with the next round of lease transfers either.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
I added tests - this caused a chain of commits since I needed to fix various things up. The commit history isn't as well-structured as I was like, but I'll put it up as is and will rely on reviewers letting me know if I should break things apart into a set of auxiliary PRs (happy to do that if it's useful).
The end result is that I have basic datadriven tests that exercise the rebalancing logic, including covering the various stopping conditions, including the traces - I did a bit of work on the log messages to be able to understand what the tests were doing.

There are a number of TODOs in this PR that I intend to understand/address before merging. I'll flag them with comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-:-OeWTtbxEtb--0Tm9HhY:b-1k4jy0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_replica_include.txt` line 33 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-OeWUXhf8caT7ms-tzw0:-OeWUXhf8caT7ms-tzw1:bnb573z) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_replica_include.txt#L33)):*
> ```help
> store-id=4 node-id=4 status=ok accepting all reported=[cpu:100, write-bandwidth:0, byte-size:0] adjusted=[cpu:100, write-bandwidth:0, byte-size:0] node-reported-cpu=100 node-adjusted-cpu=100 seq=1
> 
> # TODO(tbg,in PR): note that config= is on its own line below. This is actually
> ```

Need to make the test parser complain when it finds pieces it can't interpret, then clean up all failing tests.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_replica_include.txt` line 71 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-OeWV05-3_plySONtUwN:-OeWV05-3_plySONtUwO:betiyjy) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_replica_include.txt#L71)):*
> ```help
> # into being overloaded.
> #
> # TODO(tbg): with some refactoring, this transition could be made immediately,
> ```

Probably not worth it? cc @sumeerbhola 
Idea being that a large multiple of the grace period has passed here, so it's kind of silly to wait until the next call to actually consider store urgently overloaded.
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/mma/skewed_cpu_skewed_write_more_ranges.txt` line 41 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-OeWVE5324rzCWCHrs3Z:-OeWVE5324rzCWCHrs3_:b-vcjt5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/asim/tests/testdata/non_rand/mma/skewed_cpu_skewed_write_more_ranges.txt#L41)):*
> ```help
> cpu_util#1: last:  [s1=0.18, s2=0.18, s3=0.18, s4=0.15, s5=0.15, s6=0.15] (stddev=0.01, mean=0.17, sum=1)
> cpu_util#1: thrash_pct: [s1=9%, s2=53%, s3=53%, s4=26%, s5=26%, s6=27%]  (sum=195%)
> leases#1: first: [s1=300, s2=0, s3=0, s4=300, s5=0, s6=0] (stddev=141.42, mean=100.00, sum=600)
> ```

Need to at least review these diffs.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_lease_transfer_unbounded.txt` line 7 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-OeWUsL56xq9T5CewHp_:-OeWUsL56xq9T5CewHpa:b-3akaym) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_lease_transfer_unbounded.txt#L7)):*
> ```help
> # s1 is overloaded and local, so rebalanceStores should try and succeed to shed
> # leases from it. We disable all limits.
> # TODO(tbg,during PR): understand the frac-pending part of the "cannot add load" messages[1],
> ```

To resolve.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/allocator_state.go` line 216 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-OeWUNqV--oGRqB_ZckK:-OeWUNqV--oGRqB_ZckL:b-8ndybd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/allocator_state.go#L216)):*
> ```Go
> 
> // TODO(sumeer): temporary constants.
> // TODO(tbg): avoid direct access to this constant so that it can be configured
> ```

Would like to follow through in this PR.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/multiple_ranges.txt` line 20 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-OeWUkZC3JDVNpKWkKmC:-OeWUkZC3JDVNpKWkKmD:bg31ebh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/multiple_ranges.txt#L20)):*
> ```help
> 
> # Range 4 has the highest cpu on the leaseholder, and range 3 the highest cpu on a follower.
> store-leaseholder-msg num-top-k-replicas=2
> ```

Because 2 was the hard-coded default and I changed the default to 128 (i.e. effectively unlimited), which in this test changed the output.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
I'm ignoring the commits -- my brain is not wired to handle more than a couple of commits per PR.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> assuming the asim diffs go away.

> @sumeerbhola reviewed 1 of 1 files at r17, 1 of 1 files at r18, 2 of 2 files at r20, 17 of 17 files at r21, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-:-OeYIk9f590CPhBAqoVZ:b2q7kl0)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @tbg and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 197 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe2AiNm97ZNUGbKExg8:-OeYL3omE3v2w99l1bNo:b-qhkky4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1de6aa680a059071093b5f27a0f2a989ddafc80/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L197)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

This is `rebalanceReplicas`, which does store moves only. Leases can be shed only from the local store, so there's no loop that calls it multiple times - it's only called for the (up to one) `rebalanceStores` invocation for the localStoreID.

I made some updates (not exactly related to this comment) to the early return flow in an attempt to make it more transparent.
</blockquote></details>

This is still a bit confusing and needs a code comment.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 630 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe2CuQ0CyQBq-4Xq-w3:-OeYMRaP8y3A_DICdwjw:b5uewj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1de6aa680a059071093b5f27a0f2a989ddafc80/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L630)):*
> If we emit, say, one lease transfer but don't hit a limit, we DO want to try moving replicas, right?

I had seen some simulation results where we stopped moving more leases because the target was over a pending threshold, and then needlessly moved replicas, when waiting would allow for more lease moves. I would prefer to keep this conservative behavior which postpones replica moves from the local leaseholder until there are no leases to move.

Hopefully, that change will also eliminate the diffs in the asim output -- let's scope this PR to trying not to have asim diffs.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_replica_include.txt` line 71 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-OeWV05-3_plySONtUwN:-OeYhvNACtbi-xi243P0:b-oqahhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_replica_include.txt#L71)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Probably not worth it? cc @sumeerbhola 
Idea being that a large multiple of the grace period has passed here, so it's kind of silly to wait until the next call to actually consider store urgently overloaded.
</blockquote></details>

are you talking about `re.now.Sub(ss.overloadStartTime) < remoteStoreLeaseSheddingGraceDuration`, where that constant is 2min. Is this because we set `overloadStartTime` when attempting to rebalance, and we only call this once below. So we shouldn't need to tick here.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_replica_include.txt` line 67 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-OeYhUIy5QV5KES0Youv:-OeYhUIy5QV5KES0Youw:b-7pdone) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_replica_include.txt#L67)):*
> ```help
> store-id=4 node-id=4 status=ok accepting all reported=[cpu:100, write-bandwidth:0, byte-size:0] adjusted=[cpu:100, write-bandwidth:0, byte-size:0] node-reported-cpu=100 node-adjusted-cpu=100 seq=1
> 
> # An hour passes, so when we call rebalance-stores, s3 will transition out of
> ```

hour?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> I'm ignoring the commits -- my brain is not wired to handle more than a couple of commits per PR.

No problem if that works for you! It's surprising to me that you'd rather review a giant diff rather than isolate some of the commits, but it can get a little tangled when reviewing inside of reviewable. I addressed your comments in individual commits again - I also rearranged a bunch and pulled things ahead of the main commit. This causes a bunch of reviewable churn, but ends up with the "interesting bits" in a small number of commits at the tip of the branch. Most of the preparatory commits are now in https://github.com/cockroachdb/cockroach/pull/158182/commits, a separate PR that you should easily be able to review, since you've seen it already. (That one should definitely be good to review commit-by-commit).

I think one majorly annoying thing is that reviewable (like git) uses the commit timestamps rather than the ancestral ordering of commits, and the two are wildly divergent here. I'm running out of battery right now, but I'll post some instruction on how to look at things properly.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-:-Oe_ZerNCYX9UiTH9lhG:bvsqhtk)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @sumeerbhola and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 197 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe2AiNm97ZNUGbKExg8:-Oea34wC0seSfdDPC-cc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1de6aa680a059071093b5f27a0f2a989ddafc80/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L197)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This is still a bit confusing and needs a code comment.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 630 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe2CuQ0CyQBq-4Xq-w3:-Oea387m8LFueEHaem9y:blbfqe9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1de6aa680a059071093b5f27a0f2a989ddafc80/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L630)):*
> I had seen some simulation results where we stopped moving more leases because the target was over a pending threshold, and then needlessly moved replicas, when waiting would allow for more lease moves. I would prefer to keep this conservative behavior which postpones replica moves from the local leaseholder until there are no leases to move.

But if we hit the pending threshold, it would have skipped replica transfers - the heuristic here is that if we hit any limit, we'd skip replicas.

Anyway, I rolled this back and established the old heuristic (skip replica moves if any lease moved) - it did remove the asim diff, and we can revisit separately. I left a TODO.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_lease_transfer_unbounded.txt` line 7 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-OeWUsL56xq9T5CewHp_:-Oea8Ro68F19GCHp36OB:btnuiqm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_lease_transfer_unbounded.txt#L7)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

To resolve.
</blockquote></details>

Removed this, and instead added a comment at the log site explaining what I learned on the slack thread.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_replica_include.txt` line 33 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-OeWUXhf8caT7ms-tzw0:-Oea9ct3EaouorCtLBgI:bc142fs) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_replica_include.txt#L33)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Need to make the test parser complain when it finds pieces it can't interpret, then clean up all failing tests.
</blockquote></details>

Filed as https://github.com/cockroachdb/cockroach/issues/158191 and removed this TODO here.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_replica_include.txt` line 67 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-OeYhUIy5QV5KES0Youv:-Oea3W5576f0iX-aZqej:bputsh6) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_replica_include.txt#L67)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

hour?
</blockquote></details>

Oops, had initially ticked an hour but then that seemed excessive. Fixed.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_replica_include.txt` line 71 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-OeWV05-3_plySONtUwN:-Oea79B3BH-dR1xEW71a:b-rvjk9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/testdata/cluster_state/rebalance_stores_cpu_replica_include.txt#L71)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

are you talking about `re.now.Sub(ss.overloadStartTime) < remoteStoreLeaseSheddingGraceDuration`, where that constant is 2min. Is this because we set `overloadStartTime` when attempting to rebalance, and we only call this once below. So we shouldn't need to tick here.
</blockquote></details>

Oh, you're right. I removed this tick and kept only the second one, which I think is needed (to let the shedding period lapse). PTAL.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 24 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe_ftn92k3pdAyXCakf:-Oe_ftn92k3pdAyXCakg:bfb1eth) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L24)):*
> ```Go
> 
> // rebalanceEnv tracks the state and outcomes of a rebalanceStores invocation.
> // Recall that such an invocation is on behalf of a local store, but will
> ```

Absorbed into #158182.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 41 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe_fxbc-PjH9SzlxFhq:-Oe_fxbc-PjH9SzlxFhr:bfb1eth) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L41)):*
> ```Go
> 	// rangeMoveCount tracks the number of range moves made.
> 	rangeMoveCount int
> 	// maxRangeMoveCount is the maximum number of range moves allowed in the
> ```

Absorbed into #158182.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 46 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe_fyIP3DU1--DV3cyc:-Oe_fyIP3DU1--DV3cyd:bfb1eth) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L46)):*
> ```Go
> 	maxRangeMoveCount int
> 	// maxLeaseTransferCount is the maximum number of lease transfers allowed in
> 	// the context of the current rebalanceStores invocation. Note that because
> ```

Absorbed into #158182.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 52 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe_fz3s19p48VBN8qdB:-Oe_fz3s19p48VBN8qdC:bfb1eth) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L52)):*
> ```Go
> 	// rebalanceLeasesFromLocalStoreID invocation.
> 	maxLeaseTransferCount int
> 	// If a store's maxFractionPendingDecrease is at least this threshold, no
> ```

Absorbed into #158182.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 59 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe_g-GP0JsChBYjjgWV:-Oe_g-GP0JsChBYjjgWW:bfb1eth) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L59)):*
> ```Go
> 	// lastFailedChangeDelayDuration is the delay after a failed change before retrying.
> 	lastFailedChangeDelayDuration time.Duration
> 	// now is the timestamp representing the start time of the current
> ```

Absorbed into #158182.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 72 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe_g-m96FlsUqtzeC4P:-Oe_g-m96FlsUqtzeC4Q:bfb1eth) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L72)):*
> ```Go
> }
> 
> func newRebalanceEnv(
> ```

Absorbed into #158182.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 95 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe_g0_OA8tPo8d8cpny:-Oe_g0_OA8tPo8d8cpnz:bfb1eth) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L95)):*
> ```Go
> }
> 
> // The current usage of the multi-metric allocator has rebalanceStores called
> ```

Absorbed into #158182.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 118 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe_gzyHCAgOnZM5YwYa:-Oe_gzyHCAgOnZM5YwYb:bx8s8h7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L118)):*
> ```Go
> 	ctx context.Context, localStoreID roachpb.StoreID,
> ) []PendingRangeChange {
> 	id := re.mmaid
> ```

Absorbed the `mmaid` extraction into #158182. 
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 152 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe_hQY--KJg8ZfquGkW:-Oe_hQY--KJg8ZfquGkX:b-ipet7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L152)):*
> ```Go
> 	// via replicate_queue.go.
> 
> 	// Need deterministic order for datadriven testing.
> ```

Absorbed.
___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go` line 161 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-Oe_g3XY3g2fEQ9N0o09:-Oe_g3XY3g2fEQ9N0o0A:bfb1eth) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4e9e488e7889a4193d814150dc27646438c0b6/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_rebalance_stores.go#L161)):*
> ```Go
> 	}
> 
> 	// The caller has a fixed concurrency limit it can move ranges at, when it
> ```

Absorbed into #158182.


<!-- Sent from Reviewable.io -->

tbg committed:
mmaprototype: add 'include' test directive

tbg committed:
mmaprototype: allow multiple store load msgs per input

tbg committed:
mmaprototype: improve logging in sort[...]AndPick

tbg committed:
mmaprototype: make numtopkreplicas configurable in test

It was hard-coded to two (in the test), but I want to be able to change it for
an improved lease rebalancing test that I'm adding.

tbg committed:
mmaprototype: clarify maxRangeMoveCount and maxLeaseTransferCount

Epic: CRDB-55052

tbg committed:
mmaprototype: allow specifying limits for rebalanceStores in test

tbg reviewed (commented):
heh, I tried to rebase with

```
--exec 'GIT_COMMITTER_DATE="$(date)" git commit --amend --no-edit --allow-empty --date "$(date)"
```

in an attempt to restore sanity, but I think reviewable is too far gone at this point.

However, if you click here: https://github.com/cockroachdb/cockroach/pull/157820/commits
then scroll down to "START REVIEWING HERE" (it's an empty marker commit) and click on the title of the commit below, you can review on Github (just keep using "next" to go to the next commit). Would strongly recommend you do that, the downside being that the open comment threads are on Reviewable, so maybe check them out before actually reviewing. I think I did address all comments though.
Sorry this is such a pain!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-:-OeaBu3M4RA8ACFoq6Lw:b-5w5ph8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @sumeerbhola and @wenyihu6)



<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> @tbg dismissed @tbg from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157820)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @sumeerbhola and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/allocator_state.go` line 216 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/157820#-OeWUNqV--oGRqB_ZckK:-OeaCZWS7DIibIDn3hTm:b-wk1z3t) ([raw file](https://github.com/cockroachdb/cockroach/blob/7378d8a3cb448d1671e6de5c86771767752bb8a6/pkg/kv/kvserver/allocator/mmaprototype/allocator_state.go#L216)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Would like to follow through in this PR.
</blockquote></details>

I did this in the precursor stacked PR.


<!-- Sent from Reviewable.io -->
